File: src/infrastructure/notion_client.py
import os
import requests

from datamodel.recipe_dtos import ParsedRecipeDto


class NotionClient(object):
    page_api_url = "https://api.notion.com/v1/pages/"
    blocks_api_url = "https://api.notion.com/v1/blocks/"

    def prepare_headers(self) -> dict:
        return {
            "Authorization": f"Bearer {os.environ['NOTION_API_KEY']}",
            "Content-Type": "application/json",
            "Notion-Version": os.environ["NOTION_VERSION"],
        }

    def __prepare_children_blocks(self, recipe: ParsedRecipeDto) -> list:
        return (
            [
                {
                    "type": "heading_2",
                    "heading_2": {
                        "rich_text": [
                            {"type": "text", "text": {"content": "Ingredients"}}
                        ]
                    },
                }
            ]
            + [
                {
                    "type": "bulleted_list_item",
                    "bulleted_list_item": {
                        "rich_text": [{"type": "text", "text": {"content": ingredient}}]
                    },
                }
                for ingredient in recipe.ingredients
            ]
            + [
                {
                    "type": "heading_2",
                    "heading_2": {
                        "rich_text": [
                            {"type": "text", "text": {"content": "Instructions"}}
                        ]
                    },
                }
            ]
            + [
                {
                    "type": "numbered_list_item",
                    "numbered_list_item": {
                        "rich_text": [
                            {"type": "text", "text": {"content": instruction}}
                        ]
                    },
                }
                for instruction in recipe.instructions
            ]
        )

    def save_recipe(self, recipe: ParsedRecipeDto) -> dict:
        headers = self.prepare_headers()

        request_body = {
            "parent": {"data_source_id": os.environ["NOTION_DATABASE_ID"]},
            "properties": {
                "Name": {"title": [{"text": {"content": recipe.name}}]},
                "Source": {"url": recipe.website},
                "Tags": {"multi_select": [{"name": recipe.recipe_type}]},
            },
            "children": self.__prepare_children_blocks(recipe),
        }
        response = requests.post(self.page_api_url, headers=headers, json=request_body)

        if response.status_code != 200:
            raise Exception(f"Failed to create entry: {response.text}")

        return response.json()

    def update_recipe(self, page_id: str, recipe: ParsedRecipeDto) -> dict:
        self.clear_all_blocks(page_id)

        headers = self.prepare_headers()
        url = f"{self.blocks_api_url}{page_id}/children"

        request_body = {"children": self.__prepare_children_blocks(recipe)}
        response = requests.patch(url, headers=headers, json=request_body)

        if response.status_code != 200:
            raise Exception(f"Failed to update entry: {response.text}")

        return response.json()

    def clear_all_blocks(self, block_id: str):
        """Delete all child blocks of a given block (e.g., a page)"""
        headers = self.prepare_headers()
        url = f"{self.blocks_api_url}{block_id}/children?page_size=100"

        # Fetch existing children
        response = requests.get(url, headers=headers)
        if response.status_code != 200:
            raise Exception(f"Failed to fetch children: {response.text}")

        children = response.json().get("results", [])
        for child in children:
            child_id = child["id"]
            del_url = f"{self.blocks_api_url}{child_id}"
            del_response = requests.delete(del_url, headers=headers)
            if del_response.status_code != 200:
                raise Exception(
                    f"Failed to delete block {child_id}: {del_response.text}"
                )

File: apps/recipe-service/RecipeService.Infrastructure/Clients/NotionClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.Json.Serialization.Metadata;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using RecipeService.Domain.Entities;
using RecipeService.Domain.Enums;
using RecipeService.Infrastructure.Options;

namespace RecipeService.Infrastructure.Clients;

public class NotionClient(
    IHttpClientFactory httpClientFactory,
    IOptions<NotionClientOptions> notionClientOptions,
    ILogger<NotionClient> logger
    )
{
    private const string BaseUrl = "https://api.notion.com/v1/";
    private const string PageApiUrl = BaseUrl + "pages/";

    private void AddRequestHeaders(HttpClient httpClient)
    {
        httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {notionClientOptions.Value.ApiKey}");
        httpClient.DefaultRequestHeaders.Add("Notion-Version", notionClientOptions.Value.Version);
    }

    public async Task<JsonDocument> SaveRecipeEntityAsync(RecipeNotionEntity recipeEntity, CancellationToken cancellationToken = default)
    {
        using var httpClient = httpClientFactory.CreateClient();
        AddRequestHeaders(httpClient);

        var blocks = recipeEntity.CreateNotionBlocks();

        var requestBody = new
        {
            parent = new { data_source_id = notionClientOptions.Value.DatabaseId }, // TODO: take this value from request when ready
            properties = new
            {
                Name = new { title = new[] { new { text = new { content = recipeEntity.Name } } } },
                Source = new { url = recipeEntity.Source },
                Tags = new { multi_select = new[] { new { name = recipeEntity.RecipeType.ToDisplayString() } } },
                Processed = new { checkbox = recipeEntity.Processed }
            },
            children = blocks
        };

        var jsonOptions = new JsonSerializerOptions
        {
            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
            TypeInfoResolver = new DefaultJsonTypeInfoResolver(),
            WriteIndented = true,
        };

        logger.LogDebug("Notion API Request Body: {JsonString}", JsonSerializer.Serialize(requestBody, jsonOptions));

        var response = await httpClient.PostAsJsonAsync(PageApiUrl, requestBody, jsonOptions, cancellationToken);

        if (!response.IsSuccessStatusCode)
        {
            var errorContent = await response.Content.ReadAsStringAsync(cancellationToken);
            throw new HttpRequestException($"Failed to create entry: {errorContent}");
        }

        var responseContent = await response.Content.ReadAsStringAsync(cancellationToken);
        var jsonDocument = JsonDocument.Parse(responseContent);

        // Update the entity with the returned page ID and timestamps
        if (jsonDocument.RootElement.TryGetProperty("id", out var idProperty))
        {
            recipeEntity.PageId = idProperty.GetString();
        }

        if (jsonDocument.RootElement.TryGetProperty("created_time", out var createdTimeProperty))
        {
            if (DateTime.TryParse(createdTimeProperty.GetString(), out var createdTime))
            {
                recipeEntity.CreatedAt = createdTime;
            }
        }

        if (jsonDocument.RootElement.TryGetProperty("last_edited_time", out var lastEditedTimeProperty))
        {
            if (DateTime.TryParse(lastEditedTimeProperty.GetString(), out var lastEditedTime))
            {
                recipeEntity.LastUpdatedAt = lastEditedTime;
            }
        }

        return jsonDocument;
    }

    public async Task UpdateRecipeEntityAsync(string pageId, RecipeNotionEntity recipeEntity,
        CancellationToken cancellationToken = default)
    {
        // First create the new recipe
        var saveResponse = await SaveRecipeEntityAsync(recipeEntity, cancellationToken);

        if (!saveResponse.RootElement.TryGetProperty("id", out _))
        {
            throw new InvalidOperationException("Failed to save recipe before archiving old page.");
        }

        // Then archive the old page
        await ArchivePageAsync(pageId, cancellationToken);
    }

    private async Task ArchivePageAsync(string pageId, CancellationToken cancellationToken = default)
    {
        using var httpClient = httpClientFactory.CreateClient();
        AddRequestHeaders(httpClient);

        var url = $"{PageApiUrl}{pageId}";
        var requestBody = new { archived = true };

        var response = await httpClient.PatchAsJsonAsync(url, requestBody, cancellationToken);

        if (!response.IsSuccessStatusCode)
        {
            var errorContent = await response.Content.ReadAsStringAsync(cancellationToken);
            throw new HttpRequestException($"Failed to archive page: {errorContent}");
        }

        await response.Content.ReadAsStringAsync(cancellationToken);
    }
}
